File: go/fs.go
package jostraca

import (
	"errors"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"sync"
	"time"
)

// FS is the read+write filesystem interface used internally. Paths are
// canonical-/ at every callsite; OsFS converts to OS-native at the
// boundary via filepath.FromSlash.
type FS interface {
	ReadFile(path string) ([]byte, error)
	WriteFile(path string, data []byte) error
	Exists(path string) bool
	Stat(path string) (FileInfo, error)
	MkdirAll(path string) error
	ReadDir(path string) ([]DirEntry, error)
	Remove(path string) error
	Rename(oldpath, newpath string) error
}

// realpathFS is an optional capability: a provider that implements it lets
// the copy walk resolve a directory to its canonical path, so a symlink
// and its target compare equal and a cycle can be detected. Providers
// without it (MemFS has no symlinks) fall back to the depth cap.
type realpathFS interface {
	Realpath(path string) (string, error)
}

// chmodFS is an optional capability: a provider that implements it lets
// the atomic write-then-rename preserve an existing target's mode, which
// rename would otherwise drop along with the replaced inode. Providers
// that do not implement it simply keep the default mode.
type chmodFS interface {
	Chmod(path string, mode fs.FileMode) error
}

// exclusiveFS is an optional capability: a provider that implements it can
// create a file ONLY if it does not already exist, the way TS passes the
// `wx` flag to writeFileSync.
//
// The atomic write needs this. Checking Exists and then writing is a
// check-then-act race — another process can create the path in between —
// and, more simply, a check that ends up occupied must not fall through to
// a clobbering write. Providers without the capability get the Exists
// check plus a hard error on exhaustion, which closes the deterministic
// hole even though it cannot close the race.
type exclusiveFS interface {
	WriteFileExcl(path string, data []byte) error
}

// FileInfo is a small subset of os.FileInfo we surface across the FS
// boundary. Times are unix milliseconds for stable JSON serialisation.
type FileInfo struct {
	Name    string
	Size    int64
	Mode    fs.FileMode
	ModTime int64
	IsDir   bool
}

type DirEntry struct {
	Name  string
	IsDir bool
}

// OsFS adapts the host filesystem to the FS interface. Paths are
// converted from canonical-/ to OS-native at the boundary.
type OsFS struct{}

func (OsFS) sys(p string) string { return filepath.FromSlash(p) }

func (o OsFS) ReadFile(p string) ([]byte, error)  { return os.ReadFile(o.sys(p)) }
func (o OsFS) WriteFile(p string, b []byte) error { return os.WriteFile(o.sys(p), b, 0o644) }
func (o OsFS) Exists(p string) bool {
	_, err := os.Stat(o.sys(p))
	return err == nil
}
func (o OsFS) MkdirAll(p string) error { return os.MkdirAll(o.sys(p), 0o755) }

// WriteFileExcl implements exclusiveFS: O_EXCL fails with fs.ErrExist
// rather than truncating an existing file.
func (o OsFS) WriteFileExcl(p string, b []byte) error {
	f, err := os.OpenFile(o.sys(p), os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0o644)
	if err != nil {
		return err
	}
	if _, err := f.Write(b); err != nil {
		f.Close()
		return err
	}
	return f.Close()
}
func (o OsFS) Remove(p string) error { return os.Remove(o.sys(p)) }
func (o OsFS) Chmod(p string, mode fs.FileMode) error {
	return os.Chmod(o.sys(p), mode)
}
func (o OsFS) Realpath(p string) (string, error) {
	r, err := filepath.EvalSymlinks(o.sys(p))
	if err != nil {
		return "", err
	}
	return filepath.ToSlash(r), nil
}
func (o OsFS) Rename(a, b string) error {
	return os.Rename(o.sys(a), o.sys(b))
}
func (o OsFS) Stat(p string) (FileInfo, error) {
	fi, err := os.Stat(o.sys(p))
	if err != nil {
		return FileInfo{}, err
	}
	return FileInfo{
		Name:    fi.Name(),
		Size:    fi.Size(),
		Mode:    fi.Mode(),
		ModTime: fi.ModTime().UnixMilli(),
		IsDir:   fi.IsDir(),
	}, nil
}
func (o OsFS) ReadDir(p string) ([]DirEntry, error) {
	entries, err := os.ReadDir(o.sys(p))
	if err != nil {
		return nil, err
	}
	out := make([]DirEntry, 0, len(entries))
	for _, e := range entries {
		out = append(out, DirEntry{Name: e.Name(), IsDir: e.IsDir()})
	}
	return out, nil
}

// MemFS is an in-process filesystem backed by a string-keyed map. Safe
// for concurrent use. Paths are canonical-/ throughout.
type MemFS struct {
	mu    sync.RWMutex
	files map[string][]byte
	times map[string]int64
	dirs  map[string]bool
}

// NewMemFS constructs an empty MemFS.
func NewMemFS() *MemFS {
	return &MemFS{
		files: map[string][]byte{},
		times: map[string]int64{},
		dirs:  map[string]bool{},
	}
}

// memClean normalises a memfs path: forward slashes, no trailing /,
// no . or .. segments. Leading / is preserved (matching Node memfs's
// toJSON() output) so absolute paths like "/out/foo" are stored as
// "/out/foo", not "out/foo".
func memClean(p string) string {
	p = filepath.ToSlash(p)
	if p == "" || p == "." {
		return ""
	}
	abs := strings.HasPrefix(p, "/")
	p = strings.TrimLeft(p, "/")
	parts := []string{}
	for _, part := range strings.Split(p, "/") {
		switch part {
		case "", ".":
			continue
		case "..":
			if len(parts) > 0 {
				parts = parts[:len(parts)-1]
			}
		default:
			parts = append(parts, part)
		}
	}
	out := strings.Join(parts, "/")
	if abs {
		out = "/" + out
	}
	return out
}

func (m *MemFS) ReadFile(p string) ([]byte, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	cp := memClean(p)
	b, ok := m.files[cp]
	if !ok {
		return nil, &fs.PathError{Op: "open", Path: p, Err: os.ErrNotExist}
	}
	out := make([]byte, len(b))
	copy(out, b)
	return out, nil
}

// WriteFileExcl implements exclusiveFS: it fails rather than replacing an
// existing entry. The existence test and the store happen under one lock,
// so unlike an Exists-then-Write pair it is genuinely atomic.
func (m *MemFS) WriteFileExcl(p string, data []byte) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	cp := memClean(p)
	if _, exists := m.files[cp]; exists {
		return &os.PathError{Op: "open", Path: p, Err: fs.ErrExist}
	}
	return m.writeLocked(cp, data)
}

func (m *MemFS) WriteFile(p string, data []byte) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.writeLocked(memClean(p), data)
}

// writeLocked stores data at an already-cleaned path. Caller holds m.mu.
func (m *MemFS) writeLocked(cp string, data []byte) error {
	stored := make([]byte, len(data))
	copy(stored, data)
	m.files[cp] = stored
	m.times[cp] = time.Now().UnixMilli()
	// Implicitly create parent directories.
	if i := strings.LastIndex(cp, "/"); i > 0 {
		m.markDirsLocked(cp[:i])
	}
	return nil
}

func (m *MemFS) Exists(p string) bool {
	m.mu.RLock()
	defer m.mu.RUnlock()
	cp := memClean(p)
	if cp == "" {
		return true
	}
	if _, ok := m.files[cp]; ok {
		return true
	}
	return m.dirs[cp]
}

func (m *MemFS) Stat(p string) (FileInfo, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	cp := memClean(p)
	if cp == "" {
		return FileInfo{Name: "", IsDir: true, Mode: fs.ModeDir | 0o755}, nil
	}
	if b, ok := m.files[cp]; ok {
		return FileInfo{
			Name:    pathBase(cp),
			Size:    int64(len(b)),
			Mode:    0o644,
			ModTime: m.times[cp],
			IsDir:   false,
		}, nil
	}
	if m.dirs[cp] {
		return FileInfo{
			Name:  pathBase(cp),
			Mode:  fs.ModeDir | 0o755,
			IsDir: true,
		}, nil
	}
	return FileInfo{}, &fs.PathError{Op: "stat", Path: p, Err: os.ErrNotExist}
}

func (m *MemFS) MkdirAll(p string) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	cp := memClean(p)
	if cp == "" {
		return nil
	}
	m.markDirsLocked(cp)
	return nil
}

// markDirsLocked records every prefix of p as a directory. Caller must
// hold m.mu (write). Preserves the leading / for absolute paths so the
// stored directory keys match the form ReadDir/Stat will lookup with.
func (m *MemFS) markDirsLocked(p string) {
	if p == "" || p == "/" {
		return
	}
	abs := strings.HasPrefix(p, "/")
	inner := strings.TrimLeft(p, "/")
	if inner == "" {
		return
	}
	cur := ""
	for i, part := range strings.Split(inner, "/") {
		if i == 0 {
			if abs {
				cur = "/" + part
			} else {
				cur = part
			}
		} else {
			cur = cur + "/" + part
		}
		m.dirs[cur] = true
	}
}

func (m *MemFS) ReadDir(p string) ([]DirEntry, error) {
	m.mu.RLock()
	defer m.mu.RUnlock()
	cp := memClean(p)
	if cp != "" && !m.dirs[cp] {
		return nil, &fs.PathError{Op: "readdir", Path: p, Err: os.ErrNotExist}
	}
	prefix := cp
	if prefix != "" {
		prefix += "/"
	}
	seen := map[string]DirEntry{}
	for f := range m.files {
		if !strings.HasPrefix(f, prefix) {
			continue
		}
		rest := f[len(prefix):]
		if rest == "" {
			continue
		}
		name := rest
		isDir := false
		if i := strings.Index(rest, "/"); i >= 0 {
			name = rest[:i]
			isDir = true
		}
		if _, ok := seen[name]; !ok {
			seen[name] = DirEntry{Name: name, IsDir: isDir}
		}
	}
	for d := range m.dirs {
		if !strings.HasPrefix(d, prefix) {
			continue
		}
		rest := d[len(prefix):]
		if rest == "" {
			continue
		}
		name := rest
		if i := strings.Index(rest, "/"); i >= 0 {
			name = rest[:i]
		}
		if _, ok := seen[name]; !ok {
			seen[name] = DirEntry{Name: name, IsDir: true}
		}
	}
	names := make([]string, 0, len(seen))
	for n := range seen {
		names = append(names, n)
	}
	sort.Strings(names)
	out := make([]DirEntry, 0, len(names))
	for _, n := range names {
		out = append(out, seen[n])
	}
	return out, nil
}

func (m *MemFS) Remove(p string) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	cp := memClean(p)
	if _, ok := m.files[cp]; ok {
		delete(m.files, cp)
		delete(m.times, cp)
		return nil
	}
	if m.dirs[cp] {
		// Refuse to remove non-empty dirs to mirror os.Remove semantics.
		prefix := cp + "/"
		for f := range m.files {
			if strings.HasPrefix(f, prefix) {
				return &fs.PathError{Op: "remove", Path: p, Err: errors.New("directory not empty")}
			}
		}
		for d := range m.dirs {
			if d != cp && strings.HasPrefix(d, prefix) {
				return &fs.PathError{Op: "remove", Path: p, Err: errors.New("directory not empty")}
			}
		}
		delete(m.dirs, cp)
		return nil
	}
	return &fs.PathError{Op: "remove", Path: p, Err: os.ErrNotExist}
}

func (m *MemFS) Rename(oldpath, newpath string) error {
	m.mu.Lock()
	defer m.mu.Unlock()
	o := memClean(oldpath)
	n := memClean(newpath)
	b, ok := m.files[o]
	if !ok {
		return &fs.PathError{Op: "rename", Path: oldpath, Err: os.ErrNotExist}
	}
	delete(m.files, o)
	t := m.times[o]
	delete(m.times, o)
	m.files[n] = b
	m.times[n] = t
	if i := strings.LastIndex(n, "/"); i > 0 {
		m.markDirsLocked(n[:i])
	}
	return nil
}

// Vol returns a defensive copy of the underlying file map. Mutating the
// returned map does not affect the MemFS.
func (m *MemFS) Vol() map[string][]byte {
	m.mu.RLock()
	defer m.mu.RUnlock()
	out := make(map[string][]byte, len(m.files))
	for k, v := range m.files {
		dup := make([]byte, len(v))
		copy(dup, v)
		out[k] = dup
	}
	return out
}

func pathBase(p string) string {
	if i := strings.LastIndex(p, "/"); i >= 0 {
		return p[i+1:]
	}
	return p
}

File: rs/src/fs.rs
// fs.rs — FS trait, OsFS, and MemFS. Ported from go/fs.go.

use std::collections::HashMap;
use std::sync::RwLock;
use std::time::{SystemTime, UNIX_EPOCH};

// ─── Public types ─────────────────────────────────────────────────────────────

/// A small subset of file metadata surfaced across the FS boundary.
/// Times are unix milliseconds for stable JSON serialisation.
#[derive(Clone, Debug, Default)]
pub struct FileInfo {
    pub name:     String,
    pub size:     i64,
    pub mode:     u32,
    pub mod_time: i64,
    pub is_dir:   bool,
}

/// A directory entry.
#[derive(Clone, Debug)]
pub struct DirEntry {
    pub name:   String,
    pub is_dir: bool,
}

/// The read+write filesystem interface used internally.
pub trait FS: Send + Sync {
    fn read_file(&self, path: &str)              -> Result<Vec<u8>, String>;
    fn write_file(&self, path: &str, data: &[u8]) -> Result<(), String>;
    fn exists(&self, path: &str)                 -> bool;
    fn stat(&self, path: &str)                   -> Result<FileInfo, String>;
    fn mkdir_all(&self, path: &str)              -> Result<(), String>;
    fn read_dir(&self, path: &str)               -> Result<Vec<DirEntry>, String>;
    fn remove(&self, path: &str)                 -> Result<(), String>;
    fn rename(&self, old_path: &str, new_path: &str) -> Result<(), String>;
}

/// Optional capability: exclusive-create (fails if target exists).
pub trait ExclusiveFS: FS {
    fn write_file_excl(&self, path: &str, data: &[u8]) -> Result<(), String>;
}

/// Optional capability: chmod.
pub trait ChmodFS: FS {
    fn chmod(&self, path: &str, mode: u32) -> Result<(), String>;
}

/// Optional capability: realpath resolution.
pub trait RealpathFS: FS {
    fn realpath(&self, path: &str) -> Result<String, String>;
}

// ─── OsFS ────────────────────────────────────────────────────────────────────

/// Adapts the host filesystem to the `FS` interface.
#[derive(Clone, Default)]
pub struct OsFS;

impl OsFS {
    fn sys(&self, p: &str) -> std::path::PathBuf {
        std::path::PathBuf::from(p.replace('/', std::path::MAIN_SEPARATOR_STR))
    }
}

impl FS for OsFS {
    fn read_file(&self, path: &str) -> Result<Vec<u8>, String> {
        std::fs::read(self.sys(path)).map_err(|e| e.to_string())
    }

    fn write_file(&self, path: &str, data: &[u8]) -> Result<(), String> {
        std::fs::write(self.sys(path), data).map_err(|e| e.to_string())
    }

    fn exists(&self, path: &str) -> bool {
        self.sys(path).exists()
    }

    fn stat(&self, path: &str) -> Result<FileInfo, String> {
        let m = std::fs::metadata(self.sys(path)).map_err(|e| e.to_string())?;
        let mod_time = m.modified()
            .ok()
            .and_then(|t| t.duration_since(UNIX_EPOCH).ok())
            .map(|d| d.as_millis() as i64)
            .unwrap_or(0);
        let mode = {
            #[cfg(unix)]
            { use std::os::unix::fs::PermissionsExt; m.permissions().mode() }
            #[cfg(not(unix))]
            { if m.permissions().readonly() { 0o444 } else { 0o644 } }
        };
        Ok(FileInfo {
            name:     self.sys(path).file_name()
                          .unwrap_or_default()
                          .to_string_lossy()
                          .into_owned(),
            size:     m.len() as i64,
            mode,
            mod_time,
            is_dir:   m.is_dir(),
        })
    }

    fn mkdir_all(&self, path: &str) -> Result<(), String> {
        std::fs::create_dir_all(self.sys(path)).map_err(|e| e.to_string())
    }

    fn read_dir(&self, path: &str) -> Result<Vec<DirEntry>, String> {
        let entries = std::fs::read_dir(self.sys(path)).map_err(|e| e.to_string())?;
        let mut out = Vec::new();
        for e in entries {
            let e = e.map_err(|e| e.to_string())?;
            out.push(DirEntry {
                name:   e.file_name().to_string_lossy().into_owned(),
                is_dir: e.file_type().map(|t| t.is_dir()).unwrap_or(false),
            });
        }
        out.sort_by(|a, b| a.name.cmp(&b.name));
        Ok(out)
    }

    fn remove(&self, path: &str) -> Result<(), String> {
        let p = self.sys(path);
        if p.is_dir() {
            std::fs::remove_dir(&p).map_err(|e| e.to_string())
        } else {
            std::fs::remove_file(&p).map_err(|e| e.to_string())
        }
    }

    fn rename(&self, old_path: &str, new_path: &str) -> Result<(), String> {
        std::fs::rename(self.sys(old_path), self.sys(new_path)).map_err(|e| e.to_string())
    }
}

impl ExclusiveFS for OsFS {
    fn write_file_excl(&self, path: &str, data: &[u8]) -> Result<(), String> {
        use std::io::Write;
        let p = self.sys(path);
        let f = std::fs::OpenOptions::new()
            .write(true)
            .create_new(true)
            .open(&p)
            .map_err(|e| e.to_string())?;
        let mut f = f;
        if let Err(e) = f.write_all(data) {
            drop(f);
            let _ = std::fs::remove_file(&p);
            return Err(e.to_string());
        }
        if let Err(e) = f.flush() {
            drop(f);
            let _ = std::fs::remove_file(&p);
            return Err(e.to_string());
        }
        Ok(())
    }
}

impl ChmodFS for OsFS {
    fn chmod(&self, path: &str, mode: u32) -> Result<(), String> {
        #[cfg(unix)]
        {
            use std::os::unix::fs::PermissionsExt;
            let perms = std::fs::Permissions::from_mode(mode);
            std::fs::set_permissions(self.sys(path), perms).map_err(|e| e.to_string())
        }
        #[cfg(not(unix))]
        {
            let _ = (path, mode);
            Ok(())
        }
    }
}

impl RealpathFS for OsFS {
    fn realpath(&self, path: &str) -> Result<String, String> {
        std::fs::canonicalize(self.sys(path))
            .map(|p| p.to_string_lossy().replace('\\', "/"))
            .map_err(|e| e.to_string())
    }
}

// ─── MemFS ───────────────────────────────────────────────────────────────────

/// In-process filesystem backed by maps. Safe for concurrent use.
pub struct MemFS {
    inner: RwLock<MemFSInner>,
}

struct MemFSInner {
    files: HashMap<String, Vec<u8>>,
    times: HashMap<String, i64>,
    dirs:  HashMap<String, bool>,
}

impl MemFS {
    pub fn new() -> Self {
        MemFS {
            inner: RwLock::new(MemFSInner {
                files: HashMap::new(),
                times: HashMap::new(),
                dirs:  HashMap::new(),
            }),
        }
    }

    /// Returns a defensive copy of the underlying file map.
    pub fn vol(&self) -> HashMap<String, Vec<u8>> {
        let inner = self.inner.read().unwrap();
        inner.files.clone()
    }
}

/// Normalises a memfs path: forward slashes, no trailing /, no `.` or `..`.
pub fn mem_clean(p: &str) -> String {
    let p = p.replace('\\', "/");
    if p.is_empty() || p == "." {
        return String::new();
    }
    let abs = p.starts_with('/');
    let inner = p.trim_start_matches('/');
    let mut parts: Vec<&str> = Vec::new();
    for part in inner.split('/') {
        match part {
            "" | "." => {}
            ".." => { parts.pop(); }
            s => parts.push(s),
        }
    }
    let joined = parts.join("/");
    if abs { format!("/{}", joined) } else { joined }
}

fn now_millis() -> i64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis() as i64)
        .unwrap_or(0)
}

fn path_base(p: &str) -> &str {
    if let Some(i) = p.rfind('/') { &p[i+1..] } else { p }
}

fn mark_dirs_locked(inner: &mut MemFSInner, p: &str) {
    if p.is_empty() || p == "/" { return; }
    let abs = p.starts_with('/');
    let inner_p = p.trim_start_matches('/');
    if inner_p.is_empty() { return; }
    let mut cur = String::new();
    for (i, part) in inner_p.split('/').enumerate() {
        if i == 0 {
            cur = if abs { format!("/{}", part) } else { part.to_string() };
        } else {
            cur = format!("{}/{}", cur, part);
        }
        inner.dirs.insert(cur.clone(), true);
    }
}

impl FS for MemFS {
    fn read_file(&self, path: &str) -> Result<Vec<u8>, String> {
        let inner = self.inner.read().unwrap();
        let cp = mem_clean(path);
        inner.files.get(&cp).cloned()
            .ok_or_else(|| format!("open {}: no such file or directory", path))
    }

    fn write_file(&self, path: &str, data: &[u8]) -> Result<(), String> {
        let mut inner = self.inner.write().unwrap();
        let cp = mem_clean(path);
        inner.files.insert(cp.clone(), data.to_vec());
        inner.times.insert(cp.clone(), now_millis());
        if let Some(i) = cp.rfind('/') {
            if i > 0 { mark_dirs_locked(&mut inner, &cp[..i]); }
        }
        Ok(())
    }

    fn exists(&self, path: &str) -> bool {
        let inner = self.inner.read().unwrap();
        let cp = mem_clean(path);
        if cp.is_empty() { return true; }
        inner.files.contains_key(&cp) || inner.dirs.contains_key(&cp)
    }

    fn stat(&self, path: &str) -> Result<FileInfo, String> {
        let inner = self.inner.read().unwrap();
        let cp = mem_clean(path);
        if cp.is_empty() {
            return Ok(FileInfo { name: String::new(), is_dir: true, mode: 0o755 | 0o40000, ..Default::default() });
        }
        if let Some(data) = inner.files.get(&cp) {
            return Ok(FileInfo {
                name:     path_base(&cp).to_string(),
                size:     data.len() as i64,
                mode:     0o644,
                mod_time: *inner.times.get(&cp).unwrap_or(&0),
                is_dir:   false,
            });
        }
        if inner.dirs.contains_key(&cp) {
            return Ok(FileInfo {
                name:   path_base(&cp).to_string(),
                mode:   0o755 | 0o40000,
                is_dir: true,
                ..Default::default()
            });
        }
        Err(format!("stat {}: no such file or directory", path))
    }

    fn mkdir_all(&self, path: &str) -> Result<(), String> {
        let mut inner = self.inner.write().unwrap();
        let cp = mem_clean(path);
        if !cp.is_empty() { mark_dirs_locked(&mut inner, &cp); }
        Ok(())
    }

    fn read_dir(&self, path: &str) -> Result<Vec<DirEntry>, String> {
        let inner = self.inner.read().unwrap();
        let cp = mem_clean(path);
        if !cp.is_empty() && !inner.dirs.contains_key(&cp) {
            return Err(format!("readdir {}: no such file or directory", path));
        }
        let prefix = if cp.is_empty() { String::new() } else { format!("{}/", cp) };
        let mut seen: HashMap<String, DirEntry> = HashMap::new();
        for f in inner.files.keys() {
            if !f.starts_with(&prefix) { continue; }
            let rest = &f[prefix.len()..];
            if rest.is_empty() { continue; }
            let name = if let Some(i) = rest.find('/') { &rest[..i] } else { rest };
            let is_dir = rest.contains('/');
            seen.entry(name.to_string()).or_insert(DirEntry { name: name.to_string(), is_dir });
        }
        for d in inner.dirs.keys() {
            if !d.starts_with(&prefix) { continue; }
            let rest = &d[prefix.len()..];
            if rest.is_empty() { continue; }
            let name = if let Some(i) = rest.find('/') { &rest[..i] } else { rest };
            seen.entry(name.to_string()).or_insert(DirEntry { name: name.to_string(), is_dir: true });
        }
        let mut out: Vec<DirEntry> = seen.into_values().collect();
        out.sort_by(|a, b| a.name.cmp(&b.name));
        Ok(out)
    }

    fn remove(&self, path: &str) -> Result<(), String> {
        let mut inner = self.inner.write().unwrap();
        let cp = mem_clean(path);
        if inner.files.remove(&cp).is_some() {
            inner.times.remove(&cp);
            return Ok(());
        }
        if inner.dirs.contains_key(&cp) {
            let prefix = format!("{}/", cp);
            if inner.files.keys().any(|f| f.starts_with(&prefix)) ||
               inner.dirs.keys().any(|d| d != &cp && d.starts_with(&prefix)) {
                return Err(format!("remove {}: directory not empty", path));
            }
            inner.dirs.remove(&cp);
            return Ok(());
        }
        Err(format!("remove {}: no such file or directory", path))
    }

    fn rename(&self, old_path: &str, new_path: &str) -> Result<(), String> {
        let mut inner = self.inner.write().unwrap();
        let o = mem_clean(old_path);
        let n = mem_clean(new_path);
        let data = inner.files.remove(&o)
            .ok_or_else(|| format!("rename {}: no such file or directory", old_path))?;
        let t = inner.times.remove(&o).unwrap_or_else(now_millis);
        inner.files.insert(n.clone(), data);
        inner.times.insert(n.clone(), t);
        if let Some(i) = n.rfind('/') {
            if i > 0 { mark_dirs_locked(&mut inner, &n[..i]); }
        }
        Ok(())
    }
}

impl ExclusiveFS for MemFS {
    fn write_file_excl(&self, path: &str, data: &[u8]) -> Result<(), String> {
        let mut inner = self.inner.write().unwrap();
        let cp = mem_clean(path);
        if inner.files.contains_key(&cp) || inner.dirs.contains_key(&cp) {
            return Err(format!("open {}: file exists", path));
        }
        inner.files.insert(cp.clone(), data.to_vec());
        inner.times.insert(cp.clone(), now_millis());
        if let Some(i) = cp.rfind('/') {
            if i > 0 { mark_dirs_locked(&mut inner, &cp[..i]); }
        }
        Ok(())
    }
}
